chainntnfs/txconfnotifier.go

```diff
@@ -254,31 +309,56 @@ func (tcn *TxConfNotifier) DisconnectTip(blockHeight uint32) error {
 	tcn.currentHeight--
 	tcn.reorgDepth++
 
-	for _, txHash := range tcn.confTxsByInitialHeight[blockHeight] {
-		for _, ntfn := range tcn.confNotifications[*txHash] {
-			// If notification has been dispatched with sufficient
-			// confirmations, notify of the reversal.
-			if ntfn.dispatched {
+	for initialHeight, txHashes := range tcn.txsByInitialHeight {
+		// Now we'll go through all the transactions that were included
+		// at the same height.
+		for txHash := range txHashes {
+			for _, ntfn := range tcn.confNotifications[txHash] {
+				// First, we'll attempt to drain an update
+				// from each notification to ensure sends to the
+				// Updates channel are always non-blocking.
 				select {
-				case <-ntfn.Event.Confirmed:
-					// Drain confirmation notification instead of sending
-					// negative conf if the receiver has not processed it yet.
-					// This ensures sends to the Confirmed channel are always
-					// non-blocking.
-				case ntfn.Event.NegativeConf <- int32(tcn.reorgDepth):
+				case <-ntfn.Event.Updates:
 				case <-tcn.quit:
-					return fmt.Errorf("TxConfNotifier is exiting")
+					return errors.New("TxConfNotifier is exiting")
+				default:
 				}
-				ntfn.dispatched = false
-				continue
-			}
 
-			confHeight := blockHeight + ntfn.NumConfirmations - 1
-			ntfnSet, exists := tcn.ntfnsByConfirmHeight[confHeight]
-			if !exists {
-				continue
+				// Then, we'll check if the current transaction
+				// was included in the block currently being
+				// disconnected. If it was, we'll need to take
+				// some necessary precautions.
+				if initialHeight == blockHeight {
+					// If the transaction's confirmation notification
+					// has already been dispatched, we'll attempt to
+					// notify the client it was reorged out of the chain.
+					if ntfn.dispatched {
+						select {
+						case <-ntfn.Event.Confirmed:
+							// Drain the confirmation notification instead of sending
```

Comment:
Also should these channel sends be distinct clauses? As in _first_ drain out all the confirmed notifications, then send a negative confirmation. This would ensure that the caller has received all of what it thinks to be regular notifications, then receive the negative confirmation.

Reply:
This was already the previous behavior, I simply moved it around. I can see why it's currently implemented as is. I like your suggestion though, will fix.

Reply:
Fixed.